### PR TITLE
Move data unification

### PR DESCRIPTION
**Pull request by LOuroboros**
## Description
About 5 days ago I brought up the idea of moving Pokémon move names and descriptions from the individual arrays in which they were located to `gBattleMoves` in order to make it easier for everyone to introduce the data of entirely new moves to the codebase.
Soon after, I went and did it myself because I wanted to do it.
About 4 days ago, I got things building nicely, so after doing a bit of work and at @Bassoonian's request, I'm formally opening a draft PR for easier reviewing.

There's still a good amount of work left to do as a result of conversations that were had inside the Discord server.

- [x] We need to implement move descriptions. It was too much work to be performed manually.
- [x] We need to move the data present in `gContestMoves` to `gBattleMoves`.
- [x] We need to rename `gBattleMoves` to `gMovesInfo` or something similar to match `gSpeciesInfo`, as the array no longer contains battle data exclusively.

## **Discord contact info**
lunos4026

## **People who collaborated with me in this PR**
@Bassoonian 
@cfmnephrite
@AsparagusEduardo

### CODE REVIEW

**Comment by LOuroboros** [comment]
Oh yeah, I completely forgot. `GetMaxMoveName` and `GetZMoveName` are probably worthless now. I don't know if we want to keep them.

**Comment by LOuroboros** [comment]
Before applying a few extra tweaks that I forgot, solving conflicts and re-opening the PR, I need opinions on the latest commit that I pushed.

It saves 8 bytes of space, but it also reintroduces a few constants that I managed to remove as they were unused.

Is it worth to keep the union, or should I revert it?

**Comment by LOuroboros** [comment]
Before applying a few extra tweaks that I forgot and solving conflicts, I'd like to hear opinions on the latest commit that I pushed.

It saves 8 bytes of space, but it also reintroduces a few constants that I managed to remove as they were unused.

Is it worth to keep the union, or should I revert it?

**Comment by LOuroboros** [comment]
Aaaaaalright, thanks to @cfmnephrite's guidance I managed to fix the union. Currently it's saving 7972 bytes which is a more substantial amount than just 8.

I took the chance to perform the tweaks that I wanted to apply too.

See you all in a week or so.

File: src/rom_header_gf.c [hunk 1]
```diff
@@ -108,7 +108,6 @@ static const struct GFRomHeader sGFRomHeader = {
     //.monIconPaletteIds = gMonIconPaletteIndices,
     .monIconPalettes = gMonIconPaletteTable,
     //.monSpeciesNames = gSpeciesNames, // Handled in gSpeciesInfo
-    .moveNames = gMoveNames,
```
Comment (Bassoonian) [inline src/rom_header_gf.c:111]: We just comment out now unused fields rather than outright removing them (see other comments in this file)

**Review by Bassoonian — COMMENTED** [review]
A preliminary review. I'll make a PR with descriptions to your branch :)

File: src/battle_dynamax.c [hunk 2]
```diff
@@ -489,10 +489,7 @@ bool32 IsMaxMove(u16 move)
 // Returns the full name of a Max Move for the move usage text.
 const u8 *GetMaxMoveName(u16 move)
 {
-    if (IsMaxMove(move))
-        return gMaxMoveNames[move - FIRST_MAX_MOVE];
-    else
-        return gMaxMoveNames[0];   // Failsafe
+    return GetBattleMoveName(move);
 }
```
[the remark this answers is not in the document]
Comment (LOuroboros) [reply src/battle_dynamax.c:493]: Yeah, it does. That's what I said above:

> Oh yeah, I completely forgot. This PR makes GetMaxMoveName and GetZMoveName worthless. I don't know if we want to keep them.

I'll remove it then.

File: src/battle_z_move.c [hunk 3]
```diff
@@ -559,10 +559,7 @@ static void ZMoveSelectionDisplayMoveType(u16 zMove, u32 battler)
 
 const u8 *GetZMoveName(u16 move)
 {
-    if (IsZMove(move))
-        return gZMoveNames[move - FIRST_Z_MOVE];
-    else
-        return gZMoveNames[0];   // Failsafe
+    return GetBattleMoveName(move);
 }
```
[the remark this answers is not in the document]
Comment (LOuroboros) [reply src/battle_z_move.c:563]: Yeah, it does. That's what I said above. I'll remove it then.

File: src/data/party_menu.h [hunk 4]
```diff
@@ -719,20 +719,20 @@ struct
...
+    [MENU_FIELD_MOVES + FIELD_MOVE_SURF] = {gBattleMoves[MOVE_SURF].name.moveName, CursorCb_FieldMove},
+    [MENU_FIELD_MOVES + FIELD_MOVE_FLY] = {gBattleMoves[MOVE_FLY].name.moveName, CursorCb_FieldMove},
+    [MENU_FIELD_MOVES + FIELD_MOVE_DIVE] = {gBattleMoves[MOVE_DIVE].name.moveName, CursorCb_FieldMove},
+    [MENU_FIELD_MOVES + FIELD_MOVE_WATERFALL] = {gBattleMoves[MOVE_WATERFALL].name.moveName, CursorCb_FieldMove},
+    [MENU_FIELD_MOVES + FIELD_MOVE_TELEPORT] = {gBattleMoves[MOVE_TELEPORT].name.moveName, CursorCb_FieldMove},
+    [MENU_FIELD_MOVES + FIELD_MOVE_DIG] = {gBattleMoves[MOVE_DIG].name.moveName, CursorCb_FieldMove},
+    [MENU_FIELD_MOVES + FIELD_MOVE_SECRET_POWER] = {gBattleMoves[MOVE_SECRET_POWER].name.moveName, CursorCb_FieldMove},
+    [MENU_FIELD_MOVES + FIELD_MOVE_MILK_DRINK] = {gBattleMoves[MOVE_MILK_DRINK].name.moveName, CursorCb_FieldMove},
+    [MENU_FIELD_MOVES + FIELD_MOVE_SOFT_BOILED] = {gBattleMoves[MOVE_SOFT_BOILED].name.moveName, CursorCb_FieldMove},
+    [MENU_FIELD_MOVES + FIELD_MOVE_SWEET_SCENT] = {gBattleMoves[MOVE_SWEET_SCENT].name.moveName, CursorCb_FieldMove},
```
[the remark this answers is not in the document]
Comment (LOuroboros) [reply src/data/party_menu.h:735]: I think that I tried that but the compiler complains? Let me check again.

[the remark this answers is not in the document]
Comment (LOuroboros) [reply src/data/party_menu.h:735]: Yup, lots of errors. I'll drop a log in the Discord server.

File: src/rom_header_gf.c (hunk 1 above)
Comment (LOuroboros) [reply src/rom_header_gf.c:111]: Got it, I'll re-add it and leave it commented out then :eyes:

File: src/data/party_menu.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (LOuroboros) [reply src/data/party_menu.h:735]: > MGriffin Today at 8:26 AM
> Yeah, you can't use functions to initialize globals :(
> (Until `constexpr`, idk if C23's is even enough for that)

File: include/constants/global.h [hunk 5]
```diff
@@ -112,6 +112,7 @@
 #else
 #define MOVE_NAME_LENGTH 12
 #endif
+#define Z_MOVE_NAME_LENGTH 27
```
Comment (Bassoonian) [inline include/constants/global.h:115]: This define is unused

**Comment by LOuroboros** [comment]
Since the 3 main objectives I had with this PR had been achieved, I'd say this is ready for review.

File: include/pokemon.h [hunk 6]
```diff
@@ -442,8 +443,13 @@ struct SpeciesInfo /*0x8C*/
  /* 0x84 */ const struct FormChange *formChangeTable;
 };
 
-struct BattleMove
+struct MovesInfo
```
[the remark this answers is not in the document]
Comment (AsparagusEduardo) [reply include/pokemon.h:446]: I agree.

[the remark this answers is not in the document]
Comment (LOuroboros) [reply include/pokemon.h:446]: I don't disagree.

**Comment by AsparagusEduardo** [comment]
Oh boy, so many conflicts ^^;
Will review once solved, sorry

**Comment by LOuroboros** [comment]
> Oh boy, so many conflicts ^^; Will review once solved, sorry

Solved.

**Review by AsparagusEduardo — CHANGES_REQUESTED** [review]

File: include/pokemon.h [hunk 7]
```diff
@@ -443,8 +444,13 @@ struct SpeciesInfo /*0x8C*/
...
+struct MoveInfo
 {
+    union {
+        u8 moveName[MOVE_NAME_LENGTH + 1];
+        const u8 *longMoveName;
+    } name;
+    const u8 *description;
     u16 effect;
     u8 power;
     u8 type:5;
```
[the remark this answers is not in the document]
Comment (AsparagusEduardo) [reply include/pokemon.h:456]: As discussed on the discord, please Ignore this.

[the remark this answers is not in the document]
Comment (LOuroboros) [reply include/pokemon.h:456]: Wait, "ignore" in the sense that I should leave it as "category", or...?

File: src/data/battle_moves.h
[the remark this answers is not in the document]
Comment (LOuroboros) [reply src/data/battle_moves.h:1]: Should be "moves_info.h" imo. "moves.h" is not consistent with `src/data/pokemon/species_info.h`.

[the remark this answers is not in the document]
Comment (LOuroboros) [reply src/data/battle_moves.h:1]: Ah, but then it'd be inconsistent with this `src/data/abilities.h`.

Hmm..

File: include/pokemon.h [hunk 8]
```diff
@@ -443,8 +444,13 @@ struct SpeciesInfo /*0x8C*/
  /* 0x84 */ const struct FormChange *formChangeTable;
 };
 
-struct BattleMove
+struct MoveInfo
 {
+    union {
```
Comment (AsparagusEduardo) [inline include/pokemon.h:449]: I should've asked this before but, why aren't we using compound string for all move names?

**Review by AsparagusEduardo — CHANGES_REQUESTED** [review]
Besides the one question I have, everything looks good.

File: include/pokemon.h (hunk 8 above)
Comment (LOuroboros) [reply include/pokemon.h:449]: Compiler errors in some parts of the codebase.

Comment (AsparagusEduardo) [reply include/pokemon.h:449]: I'll see if I can PR some help your way

Comment (cfmnephrite) [reply include/pokemon.h:449]: I thought it was due to the fact that bad alignment of compound strings meant that an array was more space efficient?

Comment (LOuroboros) [reply include/pokemon.h:449]: ^ Not for me. My first priority was to get something that worked, and after I got something that worked I focused on the other aspects such as actually porting the move names and such 😂

Comment (cfmnephrite) [reply include/pokemon.h:449]: Oh, well then if we can fix those bugs it probably is worth just using a compound string for everything :P We should probably run the numbers though to check.

Comment (mrgriffin) [reply include/pokemon.h:449]: I thought Lunos (or someone else :thinking:) did run the numbers and it works out smaller to embed the names? It certainly seems like it should, we're talking 10 bytes per embedded name vs 4 bytes + (n+1) bytes for `COMPOUND_STRING`, where `n` is the length of the name. So it's pretty hard for that to get below 10 bytes unless we think most names are fewer than 6 letters long.

Comment (AsparagusEduardo) [reply include/pokemon.h:449]: I applied usage of only compound strings and ended up saving 1180 bytes.
I'm PRing to this branch those changes + description alignment in the same style as Pokédex entries.

**Review by AsparagusEduardo — APPROVED** [review]

**Pull request merged**